lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexAggregateFuture.java

```diff
@@ -0,0 +1,223 @@
+    this.resultQueue = new ArrayBlockingQueue<>(taskletIdInputMap.size());
+    this.aggOutputCodec = aggOutputCodec;
+    this.callbackHandler = callbackHandler;
+  }
+
+  /**
+   * @return the next aggregation result for the future, null if no more results.
+   */
+  public synchronized AggregateResult get() throws InterruptedException {
+    if (taskletIdInputMap.isEmpty()) {
+      return null;
+    }
+
+    return resultQueue.take();
+  }
+
+  /**
+   * @param timeout the timeout for the operation.
+   * @param timeUnit the time unit of the timeout.
+   * @return the next aggregation result for the future, within the user specified timeout, null if no more results.
+   * @throws TimeoutException if time out hits.
+   */
+  public synchronized AggregateResult get(final long timeout,
+                                          final TimeUnit timeUnit) throws InterruptedException, TimeoutException {
+    if (taskletIdInputMap.isEmpty()) {
+      return null;
+    }
+
+    final AggregateResult result = resultQueue.poll(timeout, timeUnit);
+
+    if (result == null) {
+      throw new TimeoutException();
+    }
+
+    return result;
+  }
+
+  /**
+   * @return true if there are no more results to poll.
+   */
+  public synchronized boolean isDone() {
+    return taskletIdInputMap.size() == 0;
+  }
+
+  /**
+   * A Tasklet associated with the aggregation has completed.
+   */
+  @Private
+  @Override
+  public void completed(final int taskletId, final byte[] serializedResult) {
+    try {
+      // TODO[REEF-1113]: Handle serialization failure separately in Vortex
+      final TOutput result = aggOutputCodec.decode(serializedResult);
+      removeCompletedTasklets(result, Collections.singletonList(taskletId));
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Aggregation has completed for a list of Tasklets, with an aggregated result.
+   */
+  @Private
+  @Override
+  public void aggregationCompleted(final List<Integer> pTaskletIds, final byte[] serializedResult) {
+    try {
+      // TODO[REEF-1113]: Handle serialization failure separately in Vortex
+      final TOutput result = aggOutputCodec.decode(serializedResult);
+      removeCompletedTasklets(result, pTaskletIds);
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A Tasklet associated with the aggregation has failed.
+   */
+  @Private
+  @Override
+  public void threwException(final int taskletId, final Exception exception) {
+    try {
+      removeFailedTasklets(exception, Collections.singletonList(taskletId));
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A list of Tasklets has failed during aggregation phase.
+   */
+  @Private
+  @Override
+  public void aggregationThrewException(final List<Integer> pTaskletIds, final Exception exception) {
```

Comment:
nit: I think `pTaskletIds` can be replaced by `taskletIds` since we don't have `taskletIds` field in this class.